internal/controlplane/handlers_oauth.go

```diff
@@ -231,10 +241,23 @@ func (s *Server) generateOAuthToken(ctx context.Context, provider string, code s
 		return fmt.Errorf("error exchanging code for token: %w", err)
 	}
 
+	tx, err := s.store.BeginTransaction()
+	if err != nil {
+		return status.Errorf(codes.Internal, "error starting transaction: %v", err)
+	}
+	defer s.store.Rollback(tx)
+
+	qtx := s.store.GetQuerierWithTransaction(tx)
+
+	provider, err := getOrCreateProvider(ctx, qtx, providerName, stateData)
+	if err != nil {
+		return fmt.Errorf("error getting provider: %w", err)
+	}
```

Comment:
not too keen on this, I rather have an explicit `CreateProvider` API call rather than embedding more complex logic in here.